atef/plan_utils.py

```diff
@@ -0,0 +1,242 @@
+
+        if self.allowed_devices and self.allowed_plans:
+            return self.allowed_plans, self.allowed_devices
+
+        nspace = get_default_namespace()
+        epd = existing_plans_and_devices_from_nspace(nspace=nspace)
+        self.plans_md, self.devices_md, self.plans_in_ns, self.devices_in_ns = epd
+
+        self.allowed_plans, self.allowed_devices = load_allowed_plans_and_devices(
+            existing_plans=self.plans_md, existing_devices=self.devices_md,
+            path_user_group_permissions=PERMISSIONS_PATH
+        )
+
+        return self.allowed_plans, self.allowed_devices
+
+    def register_identifier(self, identifier: str) -> None:
+        if identifier in self.run_map:
+            raise ValueError('identifier already registered')
+        self.run_map[identifier] = None
+
+
+class GlobalRunEngine:
+    # TODO: think about making this a singleton thread.
+    # Try to include (inspiration from bsqs.worker.RunEngineWorker):
+    # - .running_plan_exec_state
+    # - .re_state (property)
+    # - self._generate_continued_plan()
+    # - self.run(), self.start()
+    #   --> Used by multiprocessing.Process, QThread has different API
+    # - shutdown code?
+    RE: RunEngine
+    db: databroker.Broker
+
+    def __new__(cls):
+        if not hasattr(cls, 'instance'):
+            cls.instance = super(GlobalRunEngine, cls).__new__(cls)
+            cls.db = databroker.Broker.named('temp')
+            cls.RE = RunEngine({})
+            cls.RE.subscribe(cls.db.insert)
+
+        return cls.instance
+
+    def run_plan(
+        self,
+        state: BlueskyState,
+        item: Dict[str, Any],
+        identifier: str
+    ) -> Tuple[UUID, ...]:
+        """
+        Run a plan item, and store the resulting UUIDs in
+        ``state`` under ``identifier``
+        """
+        # separate device from component attributes
+        base, cpt = separate_attrs_from_strings(item)
+
+        # fill device, plan names with objects from namespace
+        parsed_plan = prepare_plan(
+            base,
+            plans_in_nspace=state.plans_in_ns,
+            devices_in_nspace=state.devices_in_ns,
+            allowed_plans=state.allowed_plans,
+            allowed_devices=state.allowed_devices
+        )
+
+        # reconstruct args and kwargs (operator.attrgetter(attr)(device))
+        parsed_args = get_attr_from_strings(parsed_plan['args'], cpt['args'])
+        parsed_kwargs = get_attr_from_strings(parsed_plan['kwargs'], cpt['kwargs'])
+
+        # actually run plan and store uuids
+        logger.debug(f'running plan: {parsed_plan["callable"]}'
+                     f'(*{parsed_args}, **{parsed_kwargs})')
+        run_uuids = self.RE(parsed_plan["callable"](
+                            *parsed_args, **parsed_kwargs))
+        state.run_map[identifier] = run_uuids
+        return run_uuids
+
+
+def separate_attrs_from_strings(item: Any) -> Tuple[Any, Any]:
```

Comment:
(TODO circle back because not sure what these functions do)

Reply:
Uh, I overcooked this a bit.  At some point during my testing I became convinced that the queueserver could not find components of devices, but it definitely can.  Mass deletion incoming.